src/Kestrel.Core/Internal/Http2/Http2Connection.cs

```diff
@@ -581,9 +655,167 @@ void IHttp2StreamLifetimeHandler.OnStreamCompleted(int streamId)
+                    // defined for responses MUST NOT appear in requests.
+                    throw new Http2StreamErrorException(_currentHeadersStream.StreamId, Http2ErrorCode.PROTOCOL_ERROR);
+                }
+
+                if ((_parsedPseudoHeaderFields & headerField) == headerField)
+                {
+                    // http://httpwg.org/specs/rfc7540.html#rfc.section.8.1.2.3
+                    // All HTTP/2 requests MUST include exactly one valid value for the :method, :scheme, and :path pseudo-header fields
+                    throw new Http2StreamErrorException(_currentHeadersStream.StreamId, Http2ErrorCode.PROTOCOL_ERROR);
+                }
+
+                if (headerField == PseudoHeaderFields.Method)
+                {
+                    _isMethodConnect = IsConnect(value);
+                }
+
+                _parsedPseudoHeaderFields |= headerField;
+            }
+            else if (_requestHeaderParsingState != RequestHeaderParsingState.Trailers)
+            {
+                _requestHeaderParsingState = RequestHeaderParsingState.Headers;
+            }
+
+            if (IsConnectionSpecificHeaderField(name, value))
+            {
+                throw new Http2StreamErrorException(_currentHeadersStream.StreamId, Http2ErrorCode.PROTOCOL_ERROR);
+            }
+
+            // http://httpwg.org/specs/rfc7540.html#rfc.section.8.1.2
+            // A request or response containing uppercase header field names MUST be treated as malformed (Section 8.1.2.6).
+            for (var i = 0; i < name.Length; i++)
+            {
+                if (name[i] >= 65 && name[i] <= 90)
+                {
+                    throw new Http2StreamErrorException(_currentHeadersStream.StreamId, Http2ErrorCode.PROTOCOL_ERROR);
+                }
+            }
+        }
+
+        private unsafe bool IsPseudoHeaderField(Span<byte> name, out PseudoHeaderFields headerField)
+        {
+            headerField = PseudoHeaderFields.None;
+
+            if (name.IsEmpty || name[0] != (byte)':')
+            {
+                return false;
+            }
+
+            // Skip ':'
+            name = name.Slice(1);
+
+            fixed (byte* ptr = &name.DangerousGetPinnableReference())
+            {
+                var longPtr = (ulong*)ptr;
+                var intPtr = (uint*)ptr;
+                var shortPtr = (ushort*)(ptr + 4);
+
+                switch (name.Length)
+                {
+                    case 4:
+                        if (*intPtr == _pathInt)
+                        {
+                            headerField = PseudoHeaderFields.Path;
+                        }
+                        break;
+                    case 6:
+                        if (*intPtr == _methInt && *shortPtr == _odShort)
+                        {
+                            headerField = PseudoHeaderFields.Method;
+                        }
+                        else if (*intPtr == _scheInt && *shortPtr == _meShort)
```

Comment:
This reminds me of FrameRequestHeaders.Append which is generated [here](https://github.com/aspnet/KestrelHttpServer/blob/2ce47196906a39b2821c998c109ce1aec235a806/tools/CodeGenerator/KnownHeaders.cs#L36) in KnownHeaders.

I think it might be easier to read/review/edit a code generator instead of manually bit manipulation.

Reply:
I'll leave that unoptimized for now then, since I want to prioritize finishing HTTP/2 features instead of performance work.